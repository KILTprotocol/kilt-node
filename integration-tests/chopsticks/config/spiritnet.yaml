endpoint:
 - wss://kilt-rpc.dwellir.com
db: ./db/spiritnet.db.sqlite
port: ${env.SPIRITNET_PORT}
# wasm-override: ../../target/debug/wbuild/spiritnet-runtime/spiritnet_runtime.wasm
runtime-log-level: 5

import-storage:
 TechnicalCommittee:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 Council:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 System:
  Account:
   -
    -
     - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
    - providers: 1
      data:
       free: "1000000000000000000000"
   -
    -
     - 4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ
    - providers: 1
      data:
       free: "1000000000000000000000"
