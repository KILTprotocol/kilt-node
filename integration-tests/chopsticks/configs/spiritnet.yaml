endpoint: wss://spiritnet.kilt.io
# mock-signature-host: true
db: out/spiritnet/db.sqlite
