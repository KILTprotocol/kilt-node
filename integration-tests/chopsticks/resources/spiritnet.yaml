endpoint: wss://spiritnet.kilt.io
mock-signature-host: true
db: ./db/spiritnet.db.sqlite
port: 8002

import-storage:
    technicalCommittee:
        Members: ["4pF5Y2Eo6doQHPLQj5AkndZwtomVB8ab2sRftRS2D9JDdELr"]
    council:
        Members: ["4pF5Y2Eo6doQHPLQj5AkndZwtomVB8ab2sRftRS2D9JDdELr"]
