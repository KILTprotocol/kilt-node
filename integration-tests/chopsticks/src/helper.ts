import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

export const keysAlice = keyring.addFromUri('//alice', undefined, 'ed25519')
export const keysBob = keyring.addFromUri('//bob', undefined, 'ed25519')
export const keysCharlie = keyring.addFromUri('//charlie', undefined, 'ed25519')

console.log('alice', u8aToHex(decodeAddress('4qPZ8fv6BjGoGKzfx5LtBFnEUp2b5Q5C1ErrjBNGmoFTLNHG')))
console.log('alice 2', u8aToHex(decodeAddress(keysAlice.address)))
