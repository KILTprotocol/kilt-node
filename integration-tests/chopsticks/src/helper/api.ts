/* eslint-disable @typescript-eslint/no-explicit-any */
import { ApiPromise } from '@polkadot/api'

export const xtokens = {
	parachainV2: (paraId: number) => (acc: any) => ({
		V1: {
			parents: 1,
			interior: {
				X2: [
					{ Parachain: paraId },
					{
						AccountId32: {
							network: 'Any',
							id: acc,
						},
					},
				],
			},
		},
	}),
	parachainV3: (paraId: number) => (acc: any) => ({
		V3: {
			parents: 1,
			interior: {
				X2: [
					{ Parachain: paraId },
					{
						AccountId32: {
							id: acc,
						},
					},
				],
			},
		},
	}),
	transfer:
		(token: any, dest: (dest: any) => any, weight: any = 'Unlimited') =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			api.tx.xTokens.transfer(token, amount, dest(acc), weight),
}

export const balances = {
	transferAllowDeath: ({ api }: { api: ApiPromise }, dest: string, amount: string) =>
		api.tx.balances.transferAllowDeath(dest, amount),
}

export const switchPallet = {
	switchV3:
		() =>
		({ api }: { api: ApiPromise }, acc: any, amount: string) =>
			api.tx.assetSwitchPool1.switch(amount, {
				V3: {
					parents: 0,
					interior: {
						X1: {
							AccountId32: {
								id: acc,
							},
						},
					},
				},
			}),
	switchV4:
		() =>
		({ api }: { api: ApiPromise }, acc: any, amount: string) =>
			api.tx.assetSwitchPool1.switch(amount, {
				V4: {
					parents: 0,
					interior: {
						X1: [
							{
								AccountId32: {
									id: acc,
								},
							},
						],
					},
				},
			}),
	resume:
		() =>
		({ api }: { api: ApiPromise }) =>
			api.tx.assetSwitchPool1.resumeSwitchPair(),
	pause:
		() =>
		({ api }: { api: ApiPromise }) =>
			api.tx.assetSwitchPool1.pauseSwitchPair(),
}

export const xcmPallet = {
	parachainV2: (parents: number, paraId: number) => ({
		V2: {
			parents,
			interior: {
				X1: { Parachain: paraId },
			},
		},
	}),
	parachainV3: (parents: number, paraId: any) => ({
		V3: {
			parents,
			interior: {
				X1: { Parachain: paraId },
			},
		},
	}),
	parachainV4: (parents: number, paraId: any) => ({
		V4: {
			parents,
			interior: { X1: [{ Parachain: paraId }] },
		},
	}),
	limitedTeleportAssets:
		(token: any, amount: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedTeleportAssets(
				dest,
				{
					V3: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									id: acc,
								},
							},
						},
					},
				},
				{
					V3: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	limitedReserveTransferAssetsV2:
		(token: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
				dest,
				{
					V2: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									network: 'Any',
									id: acc,
								},
							},
						},
					},
				},
				{
					V2: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	limitedReserveTransferAssetsV3:
		(token: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
				dest,
				{
					V3: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									id: acc,
								},
							},
						},
					},
				},
				{
					V3: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	limitedReserveTransferAssetsV4:
		(token: any, dest: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).limitedReserveTransferAssets(
				dest,
				{
					V4: {
						parents: 0,
						interior: {
							X1: [
								{
									AccountId32: {
										id: acc,
									},
								},
							],
						},
					},
				},
				{
					V4: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	transferAssetsV3:
		(dest: any, token: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			api.tx.polkadotXcm.transferAssets(
				dest,
				{
					V3: {
						parents: 0,
						interior: {
							X1: {
								AccountId32: {
									id: acc,
								},
							},
						},
					},
				},
				{
					V3: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	transferAssetsV4:
		(dest: any, token: any) =>
		({ api }: { api: ApiPromise }, acc: any, amount: any) =>
			api.tx.polkadotXcm.transferAssets(
				dest,
				{
					V4: {
						parents: 0,
						interior: {
							X1: [
								{
									AccountId32: {
										id: acc,
									},
								},
							],
						},
					},
				},
				{
					V4: [
						{
							id: token,
							fun: { Fungible: amount },
						},
					],
				},
				0,
				'Unlimited'
			),
	transferAssetsUsingTypeAndThenV4:
		(dest: any, token: any, xcmMessage: any) =>
		({ api }: { api: ApiPromise }, balanceToTransfer: string) =>
			api.tx.polkadotXcm.transferAssetsUsingTypeAndThen(
				dest,
				{
					V4: [
						{
							id: token,
							fun: { Fungible: balanceToTransfer },
						},
					],
				},
				'LocalReserve',
				{ V4: token },
				'LocalReserve',
				xcmMessage,
				'Unlimited'
			),
	transferAssetsUsingTypeAndThenV3:
		(dest: any, token: any, xcmMessage: any) =>
		({ api }: { api: ApiPromise }, balanceToTransfer: string) =>
			api.tx.polkadotXcm.transferAssetsUsingTypeAndThen(
				dest,
				{
					V3: [
						{
							id: token,
							fun: { Fungible: balanceToTransfer },
						},
					],
				},
				'LocalReserve',
				{ V3: token },
				'LocalReserve',
				xcmMessage,
				'Unlimited'
			),
	send:
		(destination: object) =>
		({ api }: { api: ApiPromise }, xcmMessage: object) =>
			(api.tx.xcmPallet || api.tx.polkadotXcm).send(destination, xcmMessage),
}

/**
 * Different pallets to submit tx
 */
export const tx = {
	xtokens,
	xcmPallet,
	switchPallet,
	balances,
}

/**
 * Query functions for different chains.
 * Native tokens are fetched via the system pallet, while other tokens are fetched via the tokens or assets pallet.
 *
 */
export const query = {
	balances: async ({ api }: { api: ApiPromise }, address: string) =>
		((await api.query.system.account(address)) as any).data.free.toBigInt(),
	foreignAssets:
		(assetId: any) =>
		async ({ api }: { api: ApiPromise }, address: string) => {
			const accountInfo: any = await api.query.foreignAssets.account(assetId, address)
			if (accountInfo.isNone) {
				return 0n
			}
			return accountInfo.unwrap().balance.toBigInt()
		},
	fungibles:
		(assetId: any) =>
		async ({ api }: { api: ApiPromise }, address: string) => {
			const accountInfo: any = await api.query.fungibles.account(assetId, address)
			if (accountInfo.isNone) {
				return 0n
			}
			return accountInfo.unwrap().balance.toBigInt()
		},
	tokens:
		(token: any) =>
		async ({ api }: { api: ApiPromise }, address: string) =>
			((await api.query.tokens.accounts(address, token)) as any).data.free.toBigInt(),
}
