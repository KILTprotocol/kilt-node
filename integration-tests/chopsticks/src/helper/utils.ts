import { ApiPromise, SubmittableResult } from '@polkadot/api'
import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

import type { SubmittableExtrinsic } from '@polkadot/api/types'
import type { KeyringPair } from '@polkadot/keyring/types'
import type { Codec } from '@polkadot/types/types'
import type { ExpectStatic } from 'vitest'
import { EventRecord } from '@polkadot/types/interfaces'

/**
 * Keyring instance to generate key pairs with 38 ss58Format encoding.
 */
const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

/**
 * Key pairs for Alice, Bob, and Charlie used for testing.
 */
export const keysAlice = keyring.addFromUri('//alice', undefined)
export const keysBob = keyring.addFromUri('//bob', undefined)
export const keysCharlie = keyring.addFromUri('//charlie', undefined)

/**
 * Converts a given value to a Number
 */
export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}
	return Number(value)
}

/**
 * Converts a given address to its hexadecimal representation.
 *
 * @param addr - The address to be converted. It should be a string representation of an address.
 */
export function hexAddress(addr: string) {
	return u8aToHex(decodeAddress(addr))
}

/**
 * Validates if the received balance is within the expected range considering a certain precision.
 *
 * @param previousBalance - The balance before the operation.
 * @param receivedBalance - The balance after the operation.
 * @param deltaBalance - The expected change in balance.
 * @param expect - The assertion function from the testing library.
 * @param precision - The precision of the balance validation. It must be a value between 0 and 100.
 *
 * The function calculates the expected balance by adding the deltaBalance to the previousBalance.
 * It then calculates the lower and upper bounds of the expected balance considering the precision.
 * The received balance is then checked if it falls within the range of the lower and upper bounds.
 *
 * If the precision is not between 0 and 100, the function throws an error.
 *
 * @throws {Error} If the precision is not between 0 and 100.
 */
export function validateBalanceWithPrecision(
	previousBalance: bigint,
	receivedBalance: bigint,
	deltaBalance: bigint,
	expect: ExpectStatic,
	precision: bigint
) {
	if (precision < 0n || precision > 100n) {
		throw new Error('Precision must be between 0 and 100')
	}

	const allowedError = 100n - precision
	const expectedBalance = previousBalance + deltaBalance

	const lowerBound = expectedBalance - (expectedBalance * allowedError) / 100n
	const upperBound = expectedBalance + (expectedBalance * allowedError) / 100n

	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
	expect(receivedBalance).toBeLessThanOrEqual(upperBound)
}

/**
 * Fetches the paid fees for the executed XCM message. Is only working on the sender chain.
 */
export async function getPaidXcmFees(api: ApiPromise, eventRecords: Codec[]): Promise<bigint> {
	const polkadotFees = eventRecords.find((eventRecord) =>
		api.events.polkadotXcm.FeesPaid.is((eventRecord as EventRecord).event)
	)
	if (!polkadotFees) {
		throw new Error('FeesPaid event not found')
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (polkadotFees as any).event.data.fees[0].fun.asFungible.toBigInt()
}

/**
 * Calculates the transaction fees for a given transaction.
 */
export async function calculateTxFees(
	tx: SubmittableExtrinsic<'promise', SubmittableResult>,
	account: KeyringPair
): Promise<bigint> {
	const paymentInfo = await tx.paymentInfo(account)
	return paymentInfo.partialFee.toBigInt()
}

// Constants for the balances
export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)

export const initialBalanceKILT = 100n * KILT
export const initialBalanceDOT = 100n * DOT
export const initialBalanceHDX = 100n * HDX
