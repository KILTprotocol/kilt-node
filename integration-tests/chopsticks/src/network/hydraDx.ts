import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'
import type { Config } from './types.js'
import { initialBalanceHDX, initialBalanceKILT, toNumber } from '../utils.js'

/// Options used to create the HydraDx context
export const options: SetupOption = {
	endpoint: process.env.HYDRADX_WS || ['wss://hydradx-rpc.dwellir.com', 'wss://rpc.hydradx.cloud'],
	db: './db/hydradx.db.sqlite',
	port: toNumber(process.env.HYDRADX_PORT) || 9001,
}

export const kiltTokenId = 28

/// Sets the [TechnicalCommittee] and [Council] governance to the given accounts
export function setGovernance(addr: string[]) {
	return {
		TechnicalCommittee: { Members: addr },
		Council: { Members: addr },
	}
}

/// Assigns the native tokens to an accounts
export function assignNativeTokensToAccounts(addr: string[], balance: bigint = initialBalanceHDX) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance } }]),
		},
	}
}

/// Assigns KILT tokens to an accounts
export function assignKiltTokensToAccounts(addr: string[], balance: bigint = initialBalanceKILT) {
	return {
		Tokens: {
			Accounts: addr.map((address) => [[address, kiltTokenId], { free: balance }]),
		},
	}
}

/// HydraDX ParaId
export const paraId = 2034

/// Omnipool account
export const omnipoolAccount = '7L53bUTBbfuj14UpdCNPwmgzzHSsrsTWBHX5pys32mVWM3C1'

export async function getContext(): Promise<Config> {
	return setupContext(options)
}
