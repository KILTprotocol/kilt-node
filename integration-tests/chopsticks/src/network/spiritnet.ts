import { setupContext, SetupOption } from '@acala-network/chopsticks-testing'
import type { Config } from './types.js'
import { initialBalanceKILT, toNumber } from '../utils.js'

/// Options used to create the Spiritnet context
const options: SetupOption = {
	endpoint: process.env.SPIRITNET_WS || 'wss://kilt-rpc.dwellir.com',
	db: './db/spiritnet.db.sqlite',
	port: toNumber(process.env.SPIRITNET_PORT) || 9002,
}

/// Assigns the native tokens to an accounts
export function assignNativeTokensToAccounts(addr: string[], balance: bigint = initialBalanceKILT) {
	return {
		System: {
			Account: addr.map((address) => [[address], { providers: 1, data: { free: balance } }]),
		},
	}
}

/// Sets the [technicalCommittee] and [council] governance to the given accounts
export function setGovernance(addr: string[]) {
	return {
		technicalCommittee: { Members: addr },
		council: { Members: addr },
	}
}

export function XcmPalletSafeVersion3StorageEntry() {
	return {
		polkadotXcm: {
			safeXcmVersion: 3,
		},
	}
}

/// Spiritnet ParaId
export const paraId = 2086

/// The sovereign account of HydraDx in Spiritnet
export const hydraDxSovereignAccount = '4qXPdpioJ6D8cgdeYXaukV2Y2oAQUHaX1VnGhdbSRqJn2CBt'

/// Returns the Spiritnet context for the given options
export async function getContext(): Promise<Config> {
	return setupContext(options)
}
