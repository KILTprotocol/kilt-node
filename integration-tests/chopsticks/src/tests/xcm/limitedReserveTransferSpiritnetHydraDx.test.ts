import { test } from 'vitest'
import { sendTransaction, withExpect } from '@acala-network/chopsticks-testing'

import * as SpiritnetConfig from '../../network/spiritnet.js'
import * as HydraDxConfig from '../../network/hydraDx.js'
import { keysAlice } from '../../helper.js'
import { spiritnetContext, hydradxContext, getFreeBalanceSpiritnet, getFreeBalanceHydraDxKilt } from '../index.js'

test('Limited Reserve V3 Transfers from Spiritnet Account Bob -> HydraDx', async ({ expect }) => {
	const { checkEvents, checkSystemEvents } = withExpect(expect)

	// set storage
	await spiritnetContext.dev.setStorage(SpiritnetConfig.defaultStorage(keysAlice.address))
	await hydradxContext.dev.setStorage(HydraDxConfig.defaultStorage(keysAlice.address))

	const balanceSovereignAccountHydraDxBeforeTx = await getFreeBalanceSpiritnet(HydraDxConfig.sovereignAccount)

	expect(balanceSovereignAccountHydraDxBeforeTx).eq(0)

	const balanceToTransfer = 10e12

	const signedTx = spiritnetContext.api.tx.polkadotXcm
		.limitedReserveTransferAssets(
			SpiritnetConfig.V3.hydraDxDestination,
			SpiritnetConfig.V3.hydraDxBeneficiary,
			SpiritnetConfig.V3.nativeAssetId(balanceToTransfer),
			0,
			'Unlimited'
		)
		.signAsync(keysAlice)

	const events = await sendTransaction(signedTx)

	// Produce new blocks
	// fixes api runtime disconnect warning
	await new Promise((r) => setTimeout(r, 50))
	await spiritnetContext.chain.newBlock()

	// fixes api runtime disconnect warning
	await new Promise((r) => setTimeout(r, 50))
	await hydradxContext.dev.newBlock()

	// Check events

	checkEvents(events, 'xcmpQueue').toMatchSnapshot('sender events xcm queue pallet')
	checkEvents(events, 'polkadotXcm').toMatchSnapshot('sender events xcm pallet')

	checkSystemEvents(hydradxContext, 'tokens').toMatchSnapshot('receiver events tokens')
	checkSystemEvents(hydradxContext, 'currencies').toMatchSnapshot('receiver events currencies')
	checkSystemEvents(hydradxContext, 'xcmpQueue').toMatchSnapshot('receiver events xcmpQueue')

	// check balance
	const balanceSovereignAccountHydraDxAfterTx = await getFreeBalanceSpiritnet(HydraDxConfig.sovereignAccount)
	expect(balanceSovereignAccountHydraDxAfterTx).eq(balanceToTransfer)

	let freeBalanceOmnipoolAccount = await getFreeBalanceHydraDxKilt(HydraDxConfig.omnipoolAccount)
	expect(freeBalanceOmnipoolAccount).eq(balanceToTransfer)
}, 20_000)

test('Limited Reserve V2 Transfers from Spiritnet Account Bob -> HydraDx', async ({ expect }) => {
	const { checkEvents, checkSystemEvents } = withExpect(expect)

	// Set storage
	await spiritnetContext.dev.setStorage(SpiritnetConfig.defaultStorage(keysAlice.address))
	await hydradxContext.dev.setStorage(HydraDxConfig.defaultStorage(keysAlice.address))

	const balanceSovereignAccountHydraDxBeforeTx = await getFreeBalanceSpiritnet(HydraDxConfig.sovereignAccount)

	expect(balanceSovereignAccountHydraDxBeforeTx).eq(0)

	const balanceToTransfer = 10e12

	const signedTx = spiritnetContext.api.tx.polkadotXcm
		.limitedReserveTransferAssets(
			SpiritnetConfig.V2.hydraDxDestination,
			SpiritnetConfig.V2.hydraDxBeneficiary,
			SpiritnetConfig.V2.nativeAssetId(balanceToTransfer),
			0,
			'Unlimited'
		)
		.signAsync(keysAlice)

	const events = await sendTransaction(signedTx)

	// Produce new blocks
	// fixes api runtime disconnect warning
	await new Promise((r) => setTimeout(r, 50))
	await spiritnetContext.chain.newBlock()
	// fixes api runtime disconnect warning
	await new Promise((r) => setTimeout(r, 50))
	await hydradxContext.dev.newBlock()

	// Check events

	checkEvents(events, 'xcmpQueue').toMatchSnapshot('sender events xcm queue pallet')
	checkEvents(events, 'polkadotXcm').toMatchSnapshot('sender events xcm pallet')

	checkSystemEvents(hydradxContext, 'tokens').toMatchSnapshot('receiver events tokens')
	checkSystemEvents(hydradxContext, 'currencies').toMatchSnapshot('receiver events currencies')
	checkSystemEvents(hydradxContext, 'xcmpQueue').toMatchSnapshot('receiver events xcmpQueue')

	// Check balance

	const balanceSovereignAccountHydraDxAfterTx = await getFreeBalanceSpiritnet(HydraDxConfig.sovereignAccount)
	expect(balanceSovereignAccountHydraDxAfterTx).eq(balanceToTransfer)

	let freeBalanceOmnipoolAccount = await getFreeBalanceHydraDxKilt(HydraDxConfig.omnipoolAccount)
	expect(freeBalanceOmnipoolAccount).eq(balanceToTransfer)
}, 20_000)
