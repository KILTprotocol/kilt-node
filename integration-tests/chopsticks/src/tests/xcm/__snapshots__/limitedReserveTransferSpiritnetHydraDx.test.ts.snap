// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Limited Reserve V3 Transfers from Spiritnet Account Bob -> HydraDx > receiver events currencies 1`] = `
[
  {
    "data": {
      "amount": 10000000000000,
      "currencyId": 60,
      "who": "7L1RwNWSthWi53EBAagAtVt6PUV6kfr7ofnqzv6F1YadjYaf",
    },
    "method": "Deposited",
    "section": "currencies",
  },
]
`;

exports[`Limited Reserve V3 Transfers from Spiritnet Account Bob -> HydraDx > receiver events tokens 1`] = `
[
  {
    "data": {
      "amount": 10000000000000,
      "currencyId": 60,
      "who": "7L1RwNWSthWi53EBAagAtVt6PUV6kfr7ofnqzv6F1YadjYaf",
    },
    "method": "Endowed",
    "section": "tokens",
  },
  {
    "data": {
      "amount": 10000000000000,
      "currencyId": 60,
      "who": "7L1RwNWSthWi53EBAagAtVt6PUV6kfr7ofnqzv6F1YadjYaf",
    },
    "method": "Deposited",
    "section": "tokens",
  },
]
`;

exports[`Limited Reserve V3 Transfers from Spiritnet Account Bob -> HydraDx > receiver events xcmpQueue 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
      "messageId": "(hash)",
      "weight": {
        "proofSize": 0,
        "refTime": 400000000,
      },
    },
    "method": "Success",
    "section": "xcmpQueue",
  },
]
`;

exports[`Limited Reserve V3 Transfers from Spiritnet Account Bob -> HydraDx > sender events xcm pallet 1`] = `
[
  {
    "data": {
      "outcome": {
        "Complete": {
          "proofSize": 0,
          "refTime": 400000000,
        },
      },
    },
    "method": "Attempted",
    "section": "polkadotXcm",
  },
]
`;

exports[`Limited Reserve V3 Transfers from Spiritnet Account Bob -> HydraDx > sender events xcm queue pallet 1`] = `
[
  {
    "data": {
      "messageHash": "(hash)",
    },
    "method": "XcmpMessageSent",
    "section": "xcmpQueue",
  },
]
`;
